requirements/requirements.in

```diff
@@ -107,3 +106,9 @@ csv342==1.0.0
 Faker==0.8.15
 polib==1.1.0
 werkzeug==0.11.15
+
+# Workaround for https://github.com/jazzband/pip-tools/issues/355
+# git URL installs need to be listed with -e to satisfy pip-compile
+# but once they're in requirements.txt we will strip the -e
+
+-e git+git://github.com/dimagi/architect@v0.5.7a2#egg=architect==0.5.7a2
```

Comment:
Does pip-tools add better support for versioned git URL installs? Up to this point I had not found a way to allow pinning versions of git URLs that didn't either result in a reinstall of the dependency on every pip run or possibly install the wrong version or not upgrading a dependency to the pinned version if another version was already installed. I'm not sure if I ever tried this exact syntax though, specifying the version at two places in the URL. Maybe that's the magic sauce? More context in https://github.com/dimagi/commcare-hq/pull/19560#issuecomment-367840835

To this point the best way I know to handle this situation is to avoid github URLs by pushing the library to pypi.

Reply:
The key is that there are three important places where the version is noted:
- after the `@` this URL—this is the git ref to fetch, and has nothing to do with the pip version
- after the `#egg...==` at the end of the URL—this says what version of the package this URL will satisfy
- inside the `setup.py` of the downloaded package—this is the actual version that will be recorded as the current version after install

Some comments:
- If the `#egg...==` version doesn't match the version in `setup.py` (as is the case for our architect url), it will print out a warning saying as much, and then it will use the version in `setup.py` as the current installed version of the package
- In this case where the two versions don't match, installing twice in a row will result in two installs. This makes sense because pip looks at the higher version that the URL _says_ it'll provide (`0.5.7a2`) and sees that that's different from the currently installed version (`0.5.6`, which is what's actually in the `setup.py`), and determines an install is required.
- In order to get the behavior you want (no reinstalls) you have to bump the version in `setup.py`, point the url `@` the new ref, and set the `#egg=...==` version to the same new version you bumped it to.
- If you're fine with reinstalls, you can just make sure the `#egg=...==` version in the URL is _different_ from the one in `setup.py`—that'll ensure a reinstall every time.

Reply:
Thanks, this is a useful summary.